src/XrdTpc/XrdTpcTPC.cc

```diff
@@ -119,18 +119,33 @@ int TPCHandler::opensocket_callback(void *clientp,
                                     curlsocktype purpose,
                                     struct curl_sockaddr *aInfo)
 {
-// See what kind of address will be used to connect
-//
-if (purpose == CURLSOCKTYPE_IPCXN && clientp)
-   {XrdNetAddr thePeer(&(aInfo->addr));
-    ((TPCLogRecord *)clientp)->isIPv6 =  (thePeer.isIPType(XrdNetAddrInfo::IPv6)
-                                      && !thePeer.isMapped());
-   }
+  //Return a socket file descriptor (note the clo_exec flag will be set).
+  int fd = XrdSysFD_Socket(aInfo->family, aInfo->socktype, aInfo->protocol);
+  // See what kind of address will be used to connect
+  //
+  TPCLogRecord * rec = (TPCLogRecord *)clientp;
+  if (purpose == CURLSOCKTYPE_IPCXN && clientp)
+  {XrdNetAddr thePeer(&(aInfo->addr));
+    rec->isIPv6 =  (thePeer.isIPType(XrdNetAddrInfo::IPv6)
+                    && !thePeer.isMapped());
+    // Register the socket to the packet marking manager
+    if(fd >= 0 && rec->pmarkManager) {
+      rec->pmarkManager->addFd(fd,&aInfo->addr);
+    }
+  }
 
-// Return a socket file descriptor (note the clo_exec flag will be set).
-//
-   int fd = XrdSysFD_Socket(aInfo->family, aInfo->socktype, aInfo->protocol);
-   return (fd >= 0 ? fd : CURL_SOCKET_BAD);
+  return (fd >= 0 ? fd : CURL_SOCKET_BAD);
+}
+
+int TPCHandler::closesocket_callback(void *clientp, curl_socket_t fd) {
+  TPCLogRecord * rec = (TPCLogRecord *)clientp;
+  if(rec->pmarkManager) {
+    // Destroy the PMark handle associated to the file descriptor before closing it.
+    // Otherwise, we would lose the socket usage information if the socket is closed before
+    // the PMark handle is closed.
+    rec->pmarkManager->endPmark(fd);
```

Comment:
These indirections in combination with the many `if (rec->pmarkManager)` also tell me that maybe these are better moved into `TPCLogRecord`.